node/subsystem/dispatch-gen/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use proc_macro2::TokenStream;
+use quote::{quote, ToTokens};
+use syn::{parse2, Error, Fields, FieldsNamed, FieldsUnnamed, Ident, ItemEnum, Path, Result, Type, Variant};
+use std::fmt;
+
+#[proc_macro_attribute]
+pub fn subsystem_dispatch_gen(attr: proc_macro::TokenStream, item: proc_macro::TokenStream) -> proc_macro::TokenStream {
+	let attr = proc_macro2::TokenStream::from(attr);
+	let item = proc_macro2::TokenStream::from(item);
+	proc_macro::TokenStream::from(subsystem_dispatch_gen2(attr, item))
+}
+
+/// An enum variant without base type.
+#[derive(Clone)]
+struct EnumVariantDispatchWithTy {
+	// enum ty name
+	ty: Ident,
+	// variant
+	variant: EnumVariantDispatch,
+}
+
+impl fmt::Debug for EnumVariantDispatchWithTy {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		write!(f, "{}::{:?}", self.ty, self.variant)
+	}
+}
+
+impl ToTokens for EnumVariantDispatchWithTy {
+	fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+		if let Some(inner) = &self.variant.inner {
+			let enum_name = &self.ty;
+			let variant_name = &self.variant.name;
+
+			let ts = quote! {
+				#enum_name::#variant_name(#inner::from(event))
+			};
+
+			tokens.extend(std::iter::once(ts));
+		}
+	}
+}
+
+/// An enum variant without the base type, contains the relevant inner type.
+#[derive(Clone)]
+struct EnumVariantDispatch {
+	// variant name
+	name: Ident,
+	// inner type, if any, `None` if skipped
+	// TODO make this a TokenStream and only accept verbatim
+	inner: Option<Type>,
+}
+
+impl fmt::Debug for EnumVariantDispatch {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		write!(f, "{}(..)", self.name)
+	}
+}
+
+fn prepare_enum_variant(variant: &mut Variant) -> Result<EnumVariantDispatch> {
+	let skip = variant.attrs.iter().find(|attr| attr.path.is_ident("skip")).is_some();
+	variant.attrs = variant.attrs.iter().filter(|attr| { !attr.path.is_ident("skip") }).cloned().collect::<Vec<_>>();
+
+	let variant = variant.clone();
+	let inner =
+		match variant.fields.clone() {
+			// look for one called inner
+			Fields::Named(FieldsNamed { brace_token: _, named }) if !skip => named
+				.iter()
+				.filter(|field| {
+					if let Some(ident) = &field.ident {
+						ident.to_string() == "inner".to_owned()
+					} else {
+						false
+					}
+				})
+				.map(|field| Some(field.ty.clone()))
+				.next()
+				.unwrap(),
+				// .ok_or_else(|| input.error("Missing field named `inner`"))?,
+
+			// take the first one
+			Fields::Unnamed(FieldsUnnamed { paren_token: _, unnamed }) if !skip => unnamed
+				.first()
+				.map(|field| Some(field.ty.clone())).unwrap(),
+				// .ok_or_else(|| input.error("Must at least have one inner type or be skipped"))?,
+			_ if skip => None,
+			_ => {
+				// return Err(
+				// 	input.error("Enum variant has either not skip and doesn't match anything, or somethin worse")
+				// )
+				unreachable!("fooo")
```

Comment:
Nope

Reply:
Did you see the unchecked TODO in the issue description? 🙃

Reply:
I just wanted to be sure :D